src/core/polyhedral/codegen_llvm.cc

```diff
@@ -351,6 +355,45 @@ class CodeGen_TC : public Halide::Internal::CodeGen_X86 {
           << "[LLVM-IR] After optimization:\n"
           << toString(module.get());
     }
+
+    // In a separate process call the LLC and dump the ASM
+    // Since this is for debugging purposes only atm, separate process is fine.
+    if (FLAGS_llvm_dump_asm) {
+      // Create a temporary file (do not use racy tmpname;
+      // do not use tmpnam that immediately destroys the file).
+      char fileCName[256] = "/tmp/XXXXXXXXXX";
+      auto fd = mkstemp(fileCName);
+      ScopeGuard sg([fileCName]() { unlink(fileCName); });
+      CHECK_GE(fd, 0) << "mkstemp(): " << fileCName << " returned fd: " << fd;
+      close(fd);
+
+      // Close and reopen as FILE*
+      auto tmp = fopen(fileCName, "wb+");
+      auto llvmIr = toString(module.get());
+      std::fputs(llvmIr.c_str(), tmp);
+      fclose(tmp); // Close properly so the piped command can open without race
+
+      std::string fileName(fileCName);
+      // TODO: evolve this in a better way to find LLVM
+      auto command = std::string(" `which llc || which llc-5.0` ") +
+          std::string(" -asm-verbose ") + std::string(" < ") + fileName;
+      FILE* fpipe = popen(command.c_str(), "r");
+      if (fpipe == nullptr) {
+        CHECK(false) << "Failed to popen() on command: " << command;
+      }
+      ScopeGuard sg2([fpipe]() { pclose(fpipe); });
+
+      char buffer[512];
+      std::stringstream outputss;
+      while (1) {
+        int charactersRead = fread(buffer, 1, sizeof(buffer), fpipe);
+        if (charactersRead == 0) {
+          break;
+        }
+        outputss << std::string(buffer, charactersRead);
+      }
```

Comment:
The generation of asm string should be extracted out into a separate function,
with some helper functions, e.g., for slurping in a FILE.

Reply:
Should it tough? it's a single use 12 lines of code that are properly scoped (unless I am misunderstanding which part of the code you are referring to).
I can see how the whole chunk of code under ```if (FLAGS_llvm_dump_asm) {``` could live in a separate ```getAsm``` function, so I'll start with that.

Reply:
Extracting it out (as you did) was my main point.
The other point was that I see several independent pieces in the code that communicate
with the system.  IMO, it would be better to extract those out and to make them available
at a "support" level such that they can get reused later.  Otherwise, we are likely to end
up with several similar implementations spread over the code.
Up to you.